### ISSUE TITLE
Telemetry with multiple motor ESCs, sensor conflicts

### ISSUE THREAD

**Issue by Target0815** [FrSkyRC/ETHOS-Feedback-Community#4869]
Hi Bertrand,

I would like to transfer the RPM telemetry from two motor ESCs via S.Port. There are sensor conflicts.

My question is, what needs to be changed in these ESCs so that this works without conflict?

AppID, PhyID, SensorID?

The AppID describes the type of sensor, e.g. RPM. If there are now two RPM sensors, is the AppID of the second sensor changed, or is there only one AppID with two different PhyIDs?

What is the correct way?

**Comment by bsongis** [comment]
Both AppID and PhyID must be different within the same S.Port chain.

**Comment by Target0815** [comment]
The maintainer of the ESCape32 project writes:

> There's no right or wrong here. It's implementation dependant. There's absolutely zero point in having different instance IDs with different physical IDs from the protocol perspective.
> And I will explicitly call that fix as a workaround in S.Port if it sticks because it is a workaround.
> The master always knows what physical ID it is polling, hence it shouldn't care if two nodes use the same AppID. The pair [phID,appID] is always unique for two different nodes.
> It's just silly to demand a unique AppID because it adds unnecessary complexity and hinders flexibility.

So the AppID is not changed here, only the PhyID. And IMHO he is right when he says that the combination AppID and PhyID is unique if the PhyID is different.

I don't really want to get involved in the details, but I think the S.Port documentation is in need of an overhaul if such basic things are unclear.

**Comment by neoxic** [comment]
I have the official S.Port specification in front of me which I can't publish. 

<img width="726" alt="image" src="https://github.com/user-attachments/assets/11d85397-720f-42ea-89e3-1de578e6abd2" />

It is explicitly stated that a node is distinguished on the bus by its PhID (physical ID) whereas on the application level (within one node) data values are distinguished by their AppID. The lower 4 bits of AppID are the instance ID and are used to have up to 16 values of the same type within one physical unit.

<img width="613" alt="image" src="https://github.com/user-attachments/assets/5516a332-18e5-42b5-bb54-90c37d71dcbd" />

The red portion of AppID is immutable and defines the type of data being transmitted whereas the last nibble (4 bits) is the instance ID yielding up to 16 instances of the same value type.

<img width="602" alt="image" src="https://github.com/user-attachments/assets/8eb66a43-200e-4fc6-bcf6-87f5071be9ea" />

Requiring unique AppIDs at the physical layer defies the very structure of the protocol, imposes unnecessary limitation, adds complexity (units on the bus need to be aware of each other) and hinders flexibility. The master node always knows what physical ID it's using for the current polling request, thus it _**must**_ use it as the way to discern flows of data from different units.

For example, the specification allows to have 28 * 16 = 448 identical telemetry values (RPM with AppID=0x500 for example). In other words, AppID=0x500...0x50F gives us 16 logical streams and PhID=1..28 gives us 28 physical streams, 448 in total. By requiring a unique AppID, you limit the maximum number of the same values (RPM sensors for example) to 16 _regardless of physical IDs_. That's even less than the maximum number of physical IDs (28)!

**Comment by Target0815** [comment]
In relation to my enquiry with the RPM sensor, that would be ...

lt. Bertrand:
ESC1: AppID 0x0500 PhyID 0
ESC2: AppID 0x050**1** PhyID **1**

lt. FrSky documentary (which perhaps Bertrand also wrote ...?):
ESC1: AppID 0x0500 PhyID 0
ESC2: AppID 0x0500 PhyID **1**

What is correct?

**Comment by neoxic** [comment]
@Target0815 Both variants are technically correct because there's no conflict of AppIDs within each node. These two variants should **both** work on TX.

**Comment by bsongis** [comment]
I didn't write this documentation, or at least I don't remember :)

Is it really allowed?
```
ESC1: AppID 0x0500 PhyID 0
ESC2: AppID 0x0500 PhyID 1
```

Because you should have a Sensor conflict inside Ethos ...

**Comment by neoxic** [comment]
It's in full compliance with the S.Port specification and it works with other FrSky receivers.

Moreover, FrSky sensors only allow setting up their physical ID, not application ID. Take a look:
https://www.frsky-rc.com/wp-content/uploads/2017/07/Manual/RPM.pdf

<img width="661" alt="image" src="https://github.com/user-attachments/assets/ab0b8cf1-6a1f-42b5-85aa-7e491cb2dd03" />

In other words, two identical sensors will cause a sensor conflict in Ethos despite different physical IDs. Definitely, an Ethos issue!

**Comment by Satcomix** [comment]
![17342496865872192101723248388144](https://github.com/user-attachments/assets/faab8b47-daf0-406e-8323-9ae6c28fb087)
Hello everyone,
In recent years, with FrSky telemetry sensors you have always been able to set the application ID (AppID) in addition to the sensor/physical ID (PhyID) with a USB/Wifi converter or via software and OTA/connector with transmitter in the sensor menu. Times and ETHOS have also changed, as the RPM sensor shown is from 2014, and we will soon be writing the year 2025, and in addition to S.Port, FBUS has also been supported for a long time. At that time 2014 the old D8 Mode was still used in the receivers.
regards,
Torsten

**Comment by wimalopaan** [comment]
I think this is partly a wording issue (due to the move from Version 1 to Version 2).

In Version 2 the AppID consists of two parts: the "real" AppID and a 4-bit instance number. 
Clearly, the "real" or base AppID is fixed, because a RPM sensor cannot out of a sudden morph to a GPS. The base AppID describes the nature of the sensor and the physical nature of the reported value.

So, in discussions one should clearly distinguish between (base) AppID and instance number.

But, as @neoxic already stated, there is no harm to use multiple same tuples [AppID, instance] on the bus, as long als the physID are different. The tuple [physID, AppID, instance] must be unique on the bus.

As a side note: one sensor node (one physID) may respond with different [AppId, instance] (maybe round-robin) due to the fact, that this sensor node measures different things (this is explicitly allows in the FrSky docu).

And - clearly - one physical node may respond to multiple physIDs as well.

**Comment by Target0815** [comment]
> Because you should have a Sensor conflict inside Ethos ...

I have exactly the sensor conflict ... but why? The combination of AppID and PhyID **is unique** in each case. According to the FrSky documentation, this is completely ok. Why does Ethos deviate from this?

IMHO this is a problem with Ethos. Just like the different sorting when re-reading sensors. And both problems are also related. If the sensors were generally sorted by AppID/PhyID when reading in, the list is always the same after reading in (provided no sensors are added/removed). 

As it stands, only the AppID is analysed and therefore a sensor conflict is reported. If the PhyID is taken into account, this would not happen ...

**Comment by bsongis** [comment]
@Satcomix Why do you say that AppID cannot be changed on sensors? AFAIK the AppID can be changed inside Ethos. And it can also be changed in the App if I read correctly the screenshot from @Satcomix

**Comment by bsongis** [comment]
> provided no sensors are added/removed

How do you ensure that? 

By the way I have promised to add a new function to "re-discover" sensors (which would modify them IN PLACE). But honestly if you remove the sensors, I think it is safe to remove all references to those deleted sensors in widgets / mixes / logic switchs where they are used. I don't think that what I did in opentx (nothing) is better in this point, you could have a random sensor as source of your logic switch / widget after you re-discover them!

And in all cases, it is a different problem which is not related to the sensor conflicts.

**Comment by Satcomix** [comment]
> @Satcomix Why do you say that AppID cannot be changed on sensors? AFAIK the AppID can be changed inside Ethos. And it can also be changed in the App if I read correctly the screenshot from @Satcomix

Hello Bertrand,
I think you misunderstood me a bit. Of course I have the option of changing both IDs, it's always been that way, as you can see it in the picture.
I wrote: 
In recent years, with FrSky telemetry sensors **you have always been able to set the application ID (AppID) in addition to the sensor/physical ID (PhyID)** with a USB/Wifi converter or via software and OTA/connector with transmitter in the sensor menu.
best regards,
Torsten

**Comment by Target0815** [comment]
> How do you ensure that?

The same sensors are always connected in my tests. If I read them in several times, the sorting in the telemetry list is always different. Why is this? This would not be the case if they were sorted by AppID/PhyID.

If a sensor is deleted, this sensor is clearly identifiable with AppID / InstanceID and PhyID. And precisely this sensor can then be removed from widgets / mixes / logic switches.

Why should the AppID of an RPM sensor (or other sensors) be changed? FrSky has defined which data is transmitted with the AppID 0x050, for example. The InstanceID plus PhyID is then added to distinguish between several RPM sensors. However, the AppID 0x050 always remains the same.

This is all stated in the FrSky documentation, why do you deviate from this? The handling of the sensors and these sensor conflicts are a recurring theme throughout Ethos (and previously through OpenTX). Ethos is great and can do so much, but it's things like this that are so annoying ...

**Comment by bsongis** [comment]
Sorry, my answer was not for you but for @neoxic of course!

He said this: `Moreover, FrSky sensors only allow setting up their physical ID, not application ID. Take a look:` which is plain wrong if I am not mistaken.

**Comment by Satcomix** [comment]
> Sorry, my answer was not for you but for @neoxic of course!
> 
> He said this: `Moreover, FrSky sensors only allow setting up their physical ID, not application ID. Take a look:` which is plain wrong if I am not mistaken.

Hello Bertrand,
I've been trying to explain this on Discord neoxic and wimolapaan for two days.
regards,
Torsten

**Comment by wimalopaan** [comment]
> I've been trying to explain this on Discord neoxic and wimolapaan for two days.

Have you read my comment above?
Give an example where you change the AppID (not instance) of an RPM Sebsor to an e.g. GPS.  Your screen shot above lets you choose the instance nr [0,15].

**Comment by bsongis** [comment]
@Target0815 Of course AppID and PhysID are both taken into account, otherwise Ethos would not be able to detect such a conflict.

Ethos 1.6.0 will even completely disable the check if you disable the "Sensor conflict warning" in System / Alerts (today it just drops the audio warning, but the check is done). So it will work exactly like you want.

But I have to refresh my mind and remember why it has been done this way. It has been done on purpose. I would say for more complex scenarios with redundant setups, but I need to check with the RF team.

**Comment by bsongis** [comment]
@wimalopaan Take care, the AppID is made of 2 parts, one part is not possible to change, because it describes the nature of the sensor telemetry value, and the other can be changed (from various ways, and Ethos is one of them)

**Comment by wimalopaan** [comment]
> @wimalopaan Take care, the AppID is made of 2 parts, one part is not possible to change, because it describes the nature of the sensor telemetry value, and the other can be changed (from various ways, and Ethos is one of them)

That's exactly what I wrote above.

**Comment by Target0815** [comment]
It is often unclear what "AppID" means. For the RPM sensor, the AppID is 0x050, but not 0x0500. The last digit is the InstanceID and this must be changeable, but the AppID itself is not.

**Comment by bsongis** [comment]
Indeed. But the last digit allows you to install 16 RPM sensors on your model, I think it is enough, no?

**Comment by wimalopaan** [comment]
> It is often unclear what "AppID" means. For the RPM sensor, the AppID is 0x050, but not 0x0500. The last digit is the InstanceID and this must be changeable, but the AppID itself is not.

Please read my comment above https://github.com/FrSkyRC/ETHOS-Feedback-Community/issues/4869#issuecomment-2543553864

**Comment by bsongis** [comment]
AppID is stored on 16 bits. Only the last 4 bits (16 values) can modified by the user, with Ethos (or other ways)

**Comment by bsongis** [comment]
> 
> The same sensors are always connected in my tests. If I read them in several times, the sorting in the telemetry list is always different. Why is this? This would not be the case if they were sorted by AppID/PhyID.
> 
> If a sensor is deleted, this sensor is clearly identifiable with AppID / InstanceID and PhyID. And precisely this sensor can then be removed from widgets / mixes / logic switches.
> 
> Why should the AppID of an RPM sensor (or other sensors) be changed? FrSky has defined which data is transmitted with the AppID 0x050, for example. The InstanceID plus PhyID is then added to distinguish between several RPM sensors. However, the AppID 0x050 always remains the same.
> 
> This is all stated in the FrSky documentation, why do you deviate from this? The handling of the sensors and these sensor conflicts are a recurring theme throughout Ethos (and previously through OpenTX). Ethos is great and can do so much, but it's things like this that are so annoying ...

It makes sense. But if a widget stores the AppID + PhysID of its Telemetry source, it means it will search in the list of all the sensors to find the right one (if not destroyed)? It will be slow if you have a lot of sensors and the one you need is the last one. 

Unless I introduce some caching mechanism which would be possible, but more complex!

I still think that the other way I propose (re-discover, in place) will be easier to implement, less error prone (because no reference left to destroyed sensors), and it should work perfectly.

**Comment by Satcomix** [comment]
With the SequencerX project based on oXs_RP2040 by Michel (Mstrens) I created a possibility with up to 90 telemetry sensors and 120 functions via LUA widget, **which model needs the same sensor 16 times???**, you can only have a certain number of sensors anyway, approx. 50 pieces, query/transfer, is just the bottleneck.

**Comment by bsongis** [comment]
Agreed, the bottleneck is rather on the PhysID, not AppID.

**Comment by Target0815** [comment]
> Agreed, the bottleneck is rather on the PhysID, not AppID.

Only if you see the AppID as a fixed 16 bits. However, the last 4 bits are variable without changing the AppID itself.

**Comment by bsongis** [comment]
AppID is not a fixed 16 bits. It has 12 bits fixed, plus 4 bits variable.

**Comment by Target0815** [comment]
> AppID is not a fixed 16 bits. It has 12 bits fixed, plus 4 bits variable.

I agree with you, but as you can see from the discussions, it's not the same for everyone.

**Comment by Target0815** [comment]
> It makes sense. But if a widget stores the AppID + PhysID of its Telemetry source, it means it will search in the list of all the sensors to find the right one (if not destroyed)? It will be slow if you have a lot of sensors and the one you need is the last one.

You certainly need to think a little more about what the best way is. However, if the list position is now saved in the widget after the telemetry values have been read in, this can only be a half-baked solution ... Although I'm not sure if that's really the case.

**Comment by neoxic** [comment]
> He said this: Moreover, FrSky sensors only allow setting up their physical ID, not application ID. Take a look: which is plain wrong if I am not mistaken.

@bsongis Whether instance ID can be changed on sensors or not is _**completely irrelevant**_ to the problem. @Satcomix is deliberately diverting the conversation from the real issue for some weird reason. The real issue is that by requiring unique AppIDs (16-bit numbers) on the S.Port bus, Ethos is deviating from the S.Port specification. This entails completely unnecessary limitations:
1) instead of allowing maximum 448 similar sensors as per the specs, that number is artificially limited to 16;
2) more complexity is added because now units must be able to change their instance number in order to avoid conflicts;

These two limitations can (and must) be easily avoided by simply following the specification on the Ethos side. That is exactly what other radios are doing. Is Ethos a special snowflake?

There must be **no conflicts** between sensors if their **physical IDs are different**. Period.

**Comment by Satcomix** [comment]
> > He said this: Moreover, FrSky sensors only allow setting up their physical ID, not application ID. Take a look: which is plain wrong if I am not mistaken.
> 
> @bsongis Whether instance ID can be changed on sensors or not is _**completely irrelevant**_ to the problem. @Satcomix is deliberately diverting the conversation from the real issue for some weird reason. The real issue is that by requiring unique AppIDs (16-bit numbers) on the S.Port bus, Ethos is deviating from the S.Port specification. This entails completely unnecessary limitations:
> 
>     1. instead of allowing maximum 448 similar sensors as per the specs, that number is artificially limited to 16;
> 
>     2. more complexity is added because now units must be able to change their instance number in order to avoid conflicts;
> 
> 
> These two limitations can (and must) be easily avoided by simply following the specification on the Ethos side. That is exactly what other radios are doing. Is Ethos a special snowflake?
> 
> There must be **no conflicts** between sensors if their **physical IDs are different**. Period.

Hello everyone,
If changing the InstanceID is completely irrelevant, I ask myself why you have to beat around the bush for 4 days and not just create this option, because that was the actual question. Neither more nor less.
Whether you insist out of nitpicking that it's not an ApplicationID but the 4 bits of the InstanceID that can be changed doesn't matter at all, because in the last few years I haven't found an interface anywhere that asks me to change the InstanceID, in common parlance It was always the ApplicationID, which consists of a fixed 12bit plus a variable 4bit of the instance.
I also ask myself what benefit it could have for the user of ETHOS if I can look at 448 telemetry temperature sensors, plus 448 Vario data and another 448...... What useless ballast, since the possible 16 sensors are not required, with up to 28 possible sensor/PhsicalIDs.
I'm sorry that I even asked ESCape32 on Discord this question about **adding the ability to change the ApplicationID**. It has caused more trouble than good at all, sorry for this inconvenience. 
regards,
Torsten

**Comment by Target0815** [comment]
The AppID, or the understanding that there is a fixed and a variable part of the AppID, is only a small part of the problem. Ultimately, it should be possible to change the InstanceID, i.e. the variable part of the AppID, for each sensor.

=> this is the task of the sensor manufacturer or, in this case, the ESCape32 project

It is important to avoid sensor conflicts and equally important to sort the sensor list, including the clear assignment of sensors in widgets etc.

=> that is the task of Ethos

Both tasks make the projects more rounded, sensor conflicts are simply annoying and IMHO could be eliminated quite easily if both sides are a little more flexible. However, there must also be a clear, public specification from FrSky's S.Port point of view and no documentation with an NDA requirement. Most projects with S.Port functionality copy from each other and everyone seems to add something to it ...

**Comment by bsongis-frsky** [comment]
This sensors conflict detection has been added on purpose 3 years ago. I will close the issue when I have understood myself the reason (or I will remove it completely).

In simple setups, without redundancy I do agree that the Sensor conflict detection is not needed. But it shouldn't hurt, because on a bus allowing only 28 sensors, having a limit of 16 per type is not completely wrong ;)

Meanwhile, Ethos 1.6.0 will completely disable the check of those conflicts provided the conflict warning is disabled (in System / Alerts). Ethos will allow the creation of such sensors, which is not the case in 1.5.X and before.

**Comment by Satcomix** [comment]
Hello Bertrand,
In the end it was a request from me to Senyx/ESCape32 about an expansion of the configuration interface of an ESCape32 ESC, with the possibility of setting the AppID in order to operate several ESCs at the same time with ETHOS and my FrSky transmitters/receivers with S.Port telemetry without an error message "Sensor Conflict". What is not possible in the current state with the only option of PhyID adjustment.
The discussion then started with 16bit AppID or 12bit plus 4bit InstanceID and the alleged error in the S.Port protocol (according to FrSky S.Port description 2018) of the sensor query in ETHOS. You can read the rest here.
I am still of the opinion that there is no error of the S.Port protocol in the current ETHOS, and that you always have to change the PhysID and the AppID in order to get perfect telemetry reception.
Regards,
Torsten

**Comment by Target0815** [comment]
> and that you always have to change the PhysID and the AppID in order to get perfect telemetry reception.

Bertrand has now changed this in 1.6.0 RC9 ... the AppID can be the same there.

@bsongis-frsky Are you still thinking about sorting the telemetry list? It would be good if the list was always sorted in the same way after the sensors have been read in several times.

**Comment by neoxic** [comment]
@Target0815 
> Ultimately, it should be possible to change the InstanceID, i.e. the variable part of the AppID, for each sensor. It is important to avoid sensor conflicts and equally important to sort the sensor list, including the clear assignment of sensors in widgets etc.

Application ID has nothing to do with sensor conflicts. It has been mentioned before several times. Sensor conflicts happen on the bus with two or more sensors having the same physical ID.

Using instance ID for sorting purposes causes so many questions that I'm not sure it's even worth a serious discussion. I'm sure ETHOS can come up with a much better way to sort sensors than relying on instance ID provided many commercially available S.Port sensors are simply not designed for that and other TX radios have no problem with that.

**Comment by bsongis** [comment]
@neoxic In your examples you have only one S.Port bus, but it looks too simple, no?

**Comment by bsongis** [comment]
@Target0815 Not sort the list. But replace IN-PLACE the sensors by newly created ones. Then you won't loose anything in your widgets, logic switches, etc. 

But remove all sensors will clean all references, it will remain as it is because it works perfectly!

**Comment by neoxic** [comment]
@bsongis
> In your examples you have only one S.Port bus, but it looks too simple, no?

I'm not quite sure I understand what you're asking. One S.Port bus is too simple... Should it be complicated? Should there be more than one S.Port bus?

**Comment by Target0815** [comment]
> @Target0815 Not sort the list. But replace IN-PLACE the sensors by newly created ones. Then you won't loose anything in your widgets, logic switches, etc.

Yes, I have understood. By sorting, I meant that if you read in the same sensors several times (for testing), the sensor list will never be in the same order. Surely this can be corrected ...?

**Comment by bsongis-frsky** [comment]
> I'm not quite sure I understand what you're asking. One S.Port bus is too simple... Should it be complicated? Should there be more than one S.Port bus?

Yes. The problems come when you have multiple S.Port bus. Otherwise indeed this conflict warning is not really useful. 

By the way, starting with Ethos 1.6.0 RC9 if you disable the conflict warning in System / Alerts, Ethos will behave exactly like OpenTX

**Comment by bsongis-frsky** [comment]
> Yes, I have understood. By sorting, I meant that if you read in the same sensors several times (for testing), the sensor list will never be in the same order. Surely this can be corrected ...?

By discovering IN-PLACE, the sensor list will keep its order of course!

**Comment by bsongis-frsky** [comment]
Duplicate of RF_Feedback_Group#135

**Comment by bsongis-frsky** [comment]
The issue has moved to another topic: give a way to discover IN-PLACE the telemetry sensors (keep the sensors order)

**Comment by bsongis-frsky** [comment]
An issue in the TX module has been discovered since, here the issue where it is discussed:
https://github.com/FrSkyRC/RF_Feedback_Group/issues/135
